### ISSUE TITLE
Security: Verify user ownership of browser communication pipe

### ISSUE THREAD

**Issue by taviso** [keepassxreboot/keepassxc#9393]
Hello - I asked on `#keepassxc-dev` and was told to file a security issue here.

I happened to notice the audit report announcement, but it didn't cover the browser integration at all, i.e. the scary stuff :smiley:  -- I had a quick look at how it worked, and I can see the native messaging host doesn't do any host verification, it just opens a `QLocalSocket` and assumes it's legitimate, is that intentional?

That's not reliable on Windows -- it's a global namespace, so an untrusted user can create the named pipe `\\.\pipe\org.keepassxc.KeePassXC.BrowserServer_$USER`, and when the victim starts their browser it will start negotiating with the attacker.

I understand the link is encrypted, but can't the fake server just negotiate it's own key, and then start pretending to generate passwords and so on?

I just wrote some quick c code to verify there's no test I'm missing, and as soon as my trusted user's browser starts, it connects to the untrusted server:

```
c:\Temp>whoami
desktop-bjv7vh1\untrusted
c:\Temp>pipe-test.exe
Pipe@000000EC
13832 (keepassxc-proxy.exe - desktop-bjv7vh1\Victim) Connected
Read 179, {"action":"change-public-keys","publicKey":"Cg+nY...
```

That's just `CreateNamedPipe()` with a null ACL, then `ConnectNamedPipe();ReadFile();...etc`.

I didn't look at any of the other extension code, and I didn't write a full PoC because I didn't want to implement all the crypto :smiley:

**Comment by droidmonkey** [comment]
Not sure how this is security issue, can you exfiltrate database data through this false link or just provide data? The alternative connection would not know the private key held by keepassxc or the extension so it won't be able to decrypt those messages. 

The worst case scenario we theorized with the browser extension is a fraudulent proxy application that acts as a man-in-the-middle. The fake proxy would exchange keys with the extension and exchange keys with keepassxc and perform break-and-steal in the middle.

**Comment by taviso** [comment]
Right, you can exfiltrate and/or tamper with any data the browser sends. The encryption doesn't really help here, because (as you said) the attacker can just proxy it.

Presumably you wanted to protect against that, or else why encrypt the link at all? 😃

**Comment by droidmonkey** [comment]
The browser doesn't contain any sensitive information

**Comment by taviso** [comment]
It receives generated passwords, right?

**Comment by taviso** [comment]
https://github.com/keepassxreboot/keepassxc-browser/blob/273cdfd2b87081debd42831486802831ba95c86d/keepassxc-browser/background/keepass.js#L51

Maybe I misunderstand what is happening here, is this not the browser sending passwords over the native messaging host - and therefore potentially to a malicious proxy?

**Comment by varjolintu** [comment]
> https://github.com/keepassxreboot/keepassxc-browser/blob/273cdfd2b87081debd42831486802831ba95c86d/keepassxc-browser/background/keepass.js#L51
> 
> Maybe I misunderstand what is happening here, is this not the browser sending passwords over the native messaging host - and therefore potentially to a malicious proxy?

This request is sent when credentials are saved to KeePassXC using the browser extension. In this situation the new password is needed, logically. The extension is not sending a password you are currently using back to KeePassXC at any point.

**Comment by varjolintu** [comment]
Some excellent info related to this here: https://csandker.io/2021/01/10/Offensive-Windows-IPC-1-NamedPipes.html

**Comment by varjolintu** [comment]
After giving this some thought and testing some solutions I'd suggest the following:
A new option "Restrict clients" is added to Database Settings (not a global setting, logically, so everything is stored in the database Custom Data). After enabling it user can add a full binary path and its MD5 hash to a list of verified clients. If the path or hash differs when a new connection is made to the KeePassXC's unix socket, the connection is dropped immediately. This would be disabled by default, but available for those who want to restrict any local access to their database.

The process would go something like this:
- A client connects to KeePassXC
- The process ID is retrieved using `SO_PEERCED`/`LOCAL_PEERPID` socket option
- Full path of the process is retrieved using OS related functions (e.g. `proc_pidpath()` in macOS)
- BrowserHost class checks if the restricted client list is in use, and checks the full path and MD5 hash of the client/proxy (maybe the max file size of the client/proxy could be also set just in case)

This solution would be quite simple to do, and gives some extra security. At first I was thinking something related to a separate authentication request, but storing the keys in a secure way (Keychain etc.) takes a lot of work. Some other solution related to storing an extra authentication key during `associate` to the database side could work, but client shouldn't be sending any Identification Keys to a server that pretends to be KeePassXC. So, even if Identification Keys could be retrieved some way from the client, using the proposed solution KeePassXC could restrict the connections to the socket itself.

Any thoughts?

**Comment by kraoli** [comment]
**I have a question:**
Is this only a possible when the browser integration is activated in KeepassXC? Or is the native messaging port always availible?

**And here are my thoughts, as requested by [varjolintu](https://github.com/varjolintu):**
Wouldn't it be better to enable this by default or even make it mandatory? In other words. Why to keep an potential attack vector open?

**Comment by droidmonkey** [comment]
The named pipe is only opened when browser integration is enabled.

**Comment by varjolintu** [comment]
@kraoli Enabling Client Restrictions by default could cause some problems because sometimes the proxy path is not pointing to the actual binary (e.g. in Linux with Snap/Flatpak, but we could make exceptions for those). For normal users it would be one difficult step to troubleshoot when poop hits the side of the fan.

**Comment by kraoli** [comment]
Fully understand the trouble and it might be the best possible solution. But it still gives me a bit of an uncomfortable feeling knowing this issue. But I'm not an expert in this area so my words should not matter too much.

**Comment by droidmonkey** [comment]
If you are the only user using your computer at any given time (ie, it is not a shared virtual machine or multi-user RDP host), then this is not something to worry about. The risk here is also mainly on Windows and in a multi-user environment.

**Comment by taviso** [comment]
> If you are the only user using your computer at any given time (ie, it is not a shared virtual machine or multi-user RDP host), then this is not something to worry about.

It works with a NULL ACL, so sandboxed code can spy on new passwords without a sandbox escape for example, right?

**Comment by varjolintu** [comment]
Related: https://github.com/keepassxreboot/keepassxc/pull/9779
